File: strings/longest-substring-without-repeating-chars.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        if len(s) <= 1:
            return len(s)
    
        seen_char = {}
        left = 0
        longest = 0

        for right in range(len(s)):
            current_char = s[right]
            previously_seen_char = seen_char.get(current_char)

            if previously_seen_char >= left:
                left = previously_seen_char + 1
        
            seen_char[current_char] = right
            longest = max(longest, right - left + 1)
    
        return longest

File: strings/longestSubstringwithoutRepeatingChars.js
var lengthOfLongestSubstring = function (s) {
  if (s.length <= 1) {
    return s.length;
  }

  const seenChar = {};
  let left = 0,
    longest = 0;

  for (let right = 0; right < s.length; right++) {
    const currentChar = s[right];
    const previouslySeenChar = seenChar[currentChar];

    if (previouslySeenChar >= left) {
      left = previouslySeenChar + 1;
    }
    seenChar[currentChar] = right;
    longest = Math.max(longest, right - left + 1);
  }
  return longest;
};
